src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -669,6 +669,26 @@ tx_gateway_frontend::do_abort_tm_tx(
   cluster::tm_transaction tx,
   model::timeout_clock::duration timeout,
   ss::promise<tx_errc> outcome) {
+    if (tx.status == tm_transaction::tx_status::ready) {
+        if (stm->is_actual_term(tx.etag)) {
+            // client should start a transaction before attempting to
+            // abort it. since tx has actual term we know for sure it
+            // wasn't start on a different leader
+            co_return tx_errc::timeout;
+        }
+
+        // writing ready status to overwrite an ongoing transaction if
+        // it exists on an older leader
+        auto ready_tx = co_await stm->mark_tx_ready(tx.id);
+        if (!ready_tx.has_value()) {
+            co_return tx_errc::timeout;
```

Comment:
this looks problematic. don't all the return paths need to set a value (or exception) on the `outcome` promise?

Reply:
ha-ha we found this bug in parallel, I've already fixed it when was reacting to Michal's comment

Reply:
nice. also, I think I mentioned in a separate patch, with co-routines if `co_await` receives an exceptional future then it will immediately throw, which complicates the control flow logic for the promises. maybe you'll need to run everything under a big try {} block?

Reply:
done, put it behind finally